src/vocabulary/processor.rs

```diff
@@ -0,0 +1,358 @@
+#[derive(Debug, Deserialize)]
+struct ReplaceDecoder {
+    content: String,
+    pattern: ReplacePattern,
+}
+
+impl ReplaceDecoder {
+    fn space_replacement(&self) -> Option<char> {
+        if self.content != " " {
+            return None;
+        }
+        match &self.pattern {
+            ReplacePattern::String(pattern) => {
+                let mut chars = pattern.chars();
+                let char = chars.next();
+                if let Some(replacement) = char {
+                    if chars.next().is_none() {
+                        return Some(replacement);
+                    }
+                }
+                None
+            }
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+enum ReplacePattern {
+    String(String),
+}
+
+impl TokenProcessor {
+    /// Create new `TokenProcessor` with the level defined based on tokenizer's decoders.
+    pub(crate) fn new(tokenizer: &Tokenizer) -> Result<Self> {
+        match tokenizer.get_decoder() {
+            None => Err(TokenProcessorError::UnsupportedTokenizer),
+            Some(decoder) => match decoder {
+                DecoderWrapper::ByteLevel(_) => Ok(Self {
+                    level: TokenProcessorLevel::Byte,
+                }),
+                DecoderWrapper::Sequence(decoding_sequence) => {
+                    let mut is_byte_fallback = false;
+                    let mut spacechar = ' ';
+
+                    for decoder in decoding_sequence.get_decoders() {
+                        match decoder {
+                            DecoderWrapper::ByteFallback(_) => {
+                                is_byte_fallback = true;
+                            }
+                            DecoderWrapper::Replace(replace) => {
+                                // `Replace` decoder would replace a pattern in the output with something else,
+                                // which we need to know.
+                                let decoder = Self::unpack_decoder(replace)?;
+                                if let Some(replacement) = decoder.space_replacement() {
+                                    spacechar = replacement;
+                                }
+                            }
+                            _ => {}
+                        }
+                    }
+
+                    if is_byte_fallback {
+                        Ok(Self {
+                            level: TokenProcessorLevel::ByteFallback(Mods { spacechar }),
+                        })
+                    } else {
+                        Err(TokenProcessorError::UnsupportedTokenizer)
+                    }
+                }
+                _ => Err(TokenProcessorError::UnsupportedTokenizer),
+            },
+        }
+    }
+
+    /// Operates on each token based on the level of `TokenProcessor`.
+    pub(crate) fn process(&self, token: String) -> Result<Vec<u8>> {
+        match &self.level {
+            TokenProcessorLevel::Byte => {
+                let mut bytes = vec![];
```

Comment:
I don't know whether this code is performance sensitive or not, but if it is, we should consider storing `bytes` as a field of `TokenProcessor` and `.clear()`ing/reusing it, to avoid the extra allocations.

Reply:
Sensitive part is behind `Lazy` & `static`, so we're good, here it's just a vector of bytes per each token we're currently processing, which will be saved in vocabulary. Having it as a field in `TokenProcessor` definitely is not necessary and will actually add overhead.